pkg/kapp/diff/explicit_versioned_ref.go

```diff
@@ -5,76 +5,50 @@ package diff
 
 import (
 	"fmt"
-	"reflect"
 
+	ctlres "github.com/k14s/kapp/pkg/kapp/resources"
 	"gopkg.in/yaml.v2"
 )
 
-type VersionedRefDesc struct {
-	Namespace  string `yaml:"namespace"`
-	APIVersion string `yaml:"apiVersion"`
-	Kind       string `yaml:"kind"`
-	Name       string `yaml:"name"`
-}
-
 type ExplicitVersionedRef struct {
-	Resource   VersionedResource
-	Annotation string
+	AnnotationKey string
+	Annotation    string
 }
 
 const (
 	explicitReferenceKey       = "kapp.k14s.io/versioned-explicit-ref"
 	explicitReferenceKeyPrefix = "kapp.k14s.io/versioned-explicit-ref."
 )
 
-func NewExplicitVersionedRef(res VersionedResource, annotation string) *ExplicitVersionedRef {
-	return &ExplicitVersionedRef{res, annotation}
+func NewExplicitVersionedRef(annotationKey string, annotation string) *ExplicitVersionedRef {
+	return &ExplicitVersionedRef{annotationKey, annotation}
 }
 
-// Returns true if the resource is referenced by the annotation
-func (e *ExplicitVersionedRef) IsReferenced() (bool, error) {
-	reference := VersionedRefDesc{}
-	err := yaml.Unmarshal([]byte(e.Annotation), &reference)
+func (e *ExplicitVersionedRef) AsObjectRef() (map[string]interface{}, error) {
+	var objectRef map[string]interface{}
+	err := yaml.Unmarshal([]byte(e.Annotation), &objectRef)
 	if err != nil {
-		return false, fmt.Errorf("Error unmarshalling versioned reference: %s", err)
+		return nil, fmt.Errorf("Explicit reference error: could not create object ref for explicit versioned reference")
 	}
 
-	if reference.APIVersion == "" || reference.Kind == "" || reference.Name == "" {
```

Comment:
Are there any pros of unmarshalling to a struct like :
```go
type VersionedRefDesc struct {
       Namespace  string `yaml:"namespace"`
       APIVersion string `yaml:"apiVersion"`
       Kind       string `yaml:"kind"`
       Name       string `yaml:"name"`
}
```
before unmarshalling it to a variable of type `map[string]interface{}` ?
I would prefer doing something like `reference.Name` instead of `reference["name"]`.
But I guess it won't matter. It is later unmarshalled to [coreV1.ObjectReference](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L5400) which is unlikely to change.

Reply:
I think I will add a test to keep these in check